File: 2015 ICPC North American Qualifier Contest/cuttingbrownies.py
def main():
    # Precompute win states
    win = [[False] * 501 for _ in range(501)]
    
    for i in range(1, 501):
        for j in range(1, 501):
            for k in range(1, j):
                if not win[k][i] and not win[j - k][i]:
                    win[i][j] = True
                    break

    n = int(input())
    
    for _ in range(n):
        b, d, s = input().split()
        b = int(b)
        d = int(d)
        
        print(s, end=' ')
        if s[0] == 'H':
            print("can win" if win[b][d] else "cannot win")
        else:
            print("can win" if win[d][b] else "cannot win")

if __name__ == "__main__":
    main()

File: 2015 ICPC North American Qualifier Contest/cuttingbrownies.cs
using System;

class Program {
    static bool[,] win = new bool[501, 501];
    
    static void Main() {
        // Precompute win states
        for (int i = 1; i <= 500; i++) {
            for (int j = 1; j <= 500; j++) {
                win[i, j] = false;
                
                for (int k = 1; k < j; k++) {
                    win[i, j] = win[i, j] || (!win[k, i] && !win[j - k, i]);
                }
            }
        }

        int n = int.Parse(Console.ReadLine());
        
        for (int i = 0; i < n; i++) {
            string[] parts = Console.ReadLine().Split();
            int b = int.Parse(parts[0]);
            int d = int.Parse(parts[1]);
            string s = parts[2];
            
            Console.Write(s + " ");
            if (s[0] == 'H') {
                Console.WriteLine(win[b, d] ? "can win" : "cannot win");
            } else {
                Console.WriteLine(win[d, b] ? "can win" : "cannot win");
            }
        }
    }
}